field/src/extension_field/goldilocks_field.rs

```diff
@@ -0,0 +1,474 @@
+use plonky2_util::branch_hint;
+use static_assertions::const_assert;
+
+use crate::extension_field::Extendable;
+use crate::field_types::Field64;
+use crate::goldilocks_field::{reduce128, GoldilocksField};
+
+/// Reduce the value x_lo + x_hi * 2^128 to an element in the
+/// Goldilocks field.
+#[inline(always)]
+pub(crate) fn reduce160(x_lo: u128, x_hi: u32) -> GoldilocksField {
+    // If t is in {1, ..., 2^32-1},  then t*2^128 % p == p - (t << 32)
+    let hi = <GoldilocksField as Field64>::ORDER - ((x_hi as u64) << 32);
+
+    // hi is not reduced if x_hi was 0, but that doesn't matter.
+    let (lo, cy) = x_lo.overflowing_add(hi as u128);
+    if cy {
+        // cy = true is very rare. The only way it can happen is if
+        // x_lo is at least 2^128 - (2^64 - 2^63), i.e.
+        // 0xFFFFFFFF FFFFFFFF 80000000 00000000
+        // which for randomly distributed values will only happen with
+        // probability about 2^-64.
+        branch_hint();
+        let lo = reduce128(lo).0;
+        let cy_red = <GoldilocksField as Field64>::ORDER - (1u64 << 32);
+        reduce128(lo as u128 + cy_red as u128)
+    } else {
+        reduce128(lo)
+    }
+}
+
+/*
+ * The functions extD_add_prods[0-4] are helper functions for
+ * computing products for extensions of degree D over the Goldilocks
+ * field. They are faster than the generic method because all
+ * reductions are delayed until the end which means only one per
+ * result coefficient is necessary.
+ */
+
+/// Return a, b such that a + b*2^128 = 3*x with a < 2^128 and b < 2^32.
+#[inline(always)]
+fn u128_times_3(x: u128) -> (u128, u32) {
+    let (s, cy) = x.overflowing_add(x << 1);
+    (s, (x >> 127) as u32 + cy as u32)
+}
```

Comment:
> It would be nice for the compiler to know that this is just multiplication by 3. Even though this multiplication is easily done with shifts, multiplication instructions actually ~yield better throughput on Skylake (and the same throughput on Ice Lake)~ the same throughput on Skylake and Ice Lake (and the asm is easier to get right).

Could you elaborate on what you have in mind here? I had a look at a few options, but I found calculating the overflow after using `overflowing_mul` was just as complicated as the method above.

Reply:
> Could you elaborate on what you have in mind here?

No, because I don’t actually know how to communicate the semantics to the compiler 😛 I just thought it was worth mentioning in case you had any ideas.